### PR TITLE
Graceful Task Termination

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description
At a high level, this PR introduces a new concept of "terminating states" to `TaskState` which represent that a task has begun terminating but not all drivers that were created have fully exited yet. The new states `CANCELING`, `ABORTING`, and `FAILING`) are the "terminating" equivalents to `CANCELED`, `ABORTED`, and `FAILED` which are the final state that tasks will transition into after all drivers have observed the termination signal and been destroyed.

The motivation for this change is two-fold:
1. Previously, task termination would immediately transition the task into a terminal state, which means that any in-flight drivers would not record their stats in the final `TaskInfo` response that coordinators would use to produce stage and query stats.
2. As a result of considering those tasks immediately "finished", in-flight drivers on the worker node may still be running but were not visible on the coordinator for the purposes of scheduling splits fairly across worker nodes for concurrent tasks.

## Additional context and related issues
Assumptions to note in the implementation here:
- The responsibility for when to transition from terminating to terminated is in `SqlTaskExecution` if one has been created. Otherwise, `SqlTask` will transition the state machine from terminating to termination complete (since no execution means no drivers can possibly be running).
- Existing logic that uses `TaskState#isDone()` or checks for `state == TaskState.FAILED` will now only fire when termination is fully complete. In general this behavior is still correct, but potentially slower to react than before and there may be opportunities to more eagerly react to terminating states like `TaskState.FAILING` in fault tolerant execution.
- When remote tasks fail to report full termination for a long period of time (eg: the worker is unresponsive / dead), we have to make a decision about how to proceed. This implementation will time out and fail on the coordinator after the maximum error duration has elapsed but I'm open to other implementations.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* Add CANCELING, ABORTING, and FAILING task statuses for tasks waiting on running drivers to terminate fully
```

### CODE REVIEW

**Comment by losipiuk** [comment]
Can you briefly summarize in PR description what are you working on here? :)

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java [hunk 1]
```diff
@@ -569,9 +602,16 @@ public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitioned
             return new DriverSplitRunner(this, driverContext, partitionedSplit);
         }
 
+        @Nullable
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:605]: Maybe `Optional`. Also I wonder if it is necessary to avoid creating a driver. Maybe we can simply create one and let it go the normal flow.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java [hunk 2]
```diff
@@ -569,9 +602,16 @@ public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitioned
...
         public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit partitionedSplit)
         {
+            if (taskStateMachine.getState().isTerminatingOrDone()) {
+                return null;
+            }
             Driver driver = driverFactory.createDriver(driverContext);
+            // Register driver destroyed listener to detect when termination completes
+            liveCreatedDrivers.incrementAndGet();
+            driver.getDestroyedFuture().addListener(driverAndTaskTerminationTracker.getDriverDestroyedListener(), directExecutor());
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:614]: nit: Remove `getDriverDestroyedListener` and pass the `driverAndTaskTerminationTracker` directly

**Review by arhimondr — COMMENTED** [review]
The implementation looks good to me. Though the code related to task state transitioning is rather complex and it is easy to miss something. 

I wonder how can we better test this change to ensure there's no regression.

One thing that I can think of is maybe run some number of small limit queries in a loop (which trigger "cancel") and see if there's no visible side effects (no latency increase, no abandoned tasks are left behind, final query info is consistent).

Similar can be tested with query cancellation? Maybe it makes sense to run some number of queries in a loop cancelling them midway and then ensure there's no abandoned tasks / splits are left behind.

**Comment by arhimondr** [comment]
`plugin/trino-delta-lake`, `plugin/trino-iceberg` timed out. It can be a fluke, but there's a chance that queries started running longer due to the change. We should probably check that (e.g.: maybe compare an average run time before and after this change).

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 2 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java]: Done

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 1 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:605]: The only usage site is in `DriverSplitRunner` where `null` is cleaner to handle than `Optional`, so I'd prefer to keep this `@Nullable`.

Sure, we could create the driver anyway- but there's still the possibility that `DriverFactory#noMoreDrivers()` is called concurrently so we could fail to create the driver. Actually, I'll go ahead and update the logic to avoid the `taskStateMachine.getState().isTerminatingOrDone()` check but handle `DriverFactory#createDriver(DriverContext)` failing since that's the core real root of what we need to handle here by returning null.

File: core/trino-main/src/main/java/io/trino/execution/SqlTask.java [hunk 3]
```diff
@@ -145,7 +151,7 @@ private SqlTask(
                 // Pass a memory context supplier instead of a memory context to the output buffer,
                 // because we haven't created the task context that holds the memory context yet.
                 () -> queryContext.getTaskContextByTaskId(taskId).localMemoryContext(),
-                () -> notifyStatusChanged(),
+                this::notifyStatusChanged,
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/SqlTask.java:154]: Unrelated change. Move to a separate commit.

File: testing/trino-testing/src/main/java/io/trino/testing/QueryAssertions.java [hunk 4]
```diff
@@ -94,6 +100,51 @@ else if (count.isPresent()) {
         }
     }
 
+    private static void assertDistributedUpdate(DistributedQueryRunner distributedQueryRunner, Session session, @Language("SQL") String sql, OptionalLong count, Optional<Consumer<Plan>> planAssertion)
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply testing/trino-testing/src/main/java/io/trino/testing/QueryAssertions.java:103]: The failure messages include the queryId so that the failing execution can be correlated to log messages that may be related.

File: core/trino-main/src/main/java/io/trino/execution/SqlTask.java (hunk 3 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTask.java:156]: Sure, done.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java [hunk 5]
```diff
@@ -108,6 +110,8 @@
 
     // number of created PrioritizedSplitRunners that haven't yet finished
     private final AtomicLong remainingSplitRunners = new AtomicLong();
+    // number of created Drivers that haven't yet been destroyed
+    private final AtomicLong liveCreatedDrivers = new AtomicLong();
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:114]: nit: I would recommend encapsulating this counter in `DriverAndTaskTerminationTracker` and change the methods in `DriverAndTaskTerminationTracker`. For example:

- `DriverAndTaskTerminationTracker#driverCreated`
- `DriverAndTaskTerminationTracker#driverDestroyed`
- `DriverAndTaskTerminationTracker#getLiveDrivers` (for stats)
- `DriverAndTaskTerminationTracker#attemptTaskTermination` (or something along the lines)

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java [hunk 6]
```diff
@@ -93,16 +93,20 @@ public List<OperatorFactory> getOperatorFactories()
...
-                operators.add(operator);
+            synchronized (this) {
+                // must check noMoreDrivers after acquiring the lock
+                checkState(!noMoreDrivers, "noMoreDrivers is already set");
+                for (OperatorFactory operatorFactory : operatorFactories) {
+                    Operator operator = operatorFactory.createOperator(driverContext);
+                    operators.add(operator);
+                }
             }
+            // Driver creation can continue without holding the lock
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:109]: Could you please explain why this change is needed? Is this just an efficiency performance? Or was there a deadlock?

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java [hunk 7]
```diff
@@ -768,84 +782,113 @@ private synchronized SplitAssignment getSplitAssignment(PlanNodeId planNodeId, i
...
+            if (request != null) {
+                request.cancel(true);
+            }
 
-    @Override
-    public synchronized void abort()
-    {
-        if (getTaskStatus().getState().isDone()) {
-            return;
+            taskStatusFetcher.stop();
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:846]: Why doing it twice? Please add a comment if it is not a mistake

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java [hunk 8]
```diff
@@ -920,25 +968,34 @@ public void onFailure(Throwable t)
...
+                // we have to set the local query info directly so that we stop trying to fetch
+                // updated TaskInfo from the worker. This way, the task on the worker eventually
+                // expires due to lack of activity.
+                List<ExecutionFailureInfo> failures = ImmutableList.<ExecutionFailureInfo>builderWithExpectedSize(taskStatus.getFailures().size() + 1)
+                        .add(toFailure(cause))
+                        .addAll(taskStatus.getFailures())
+                        .build();
+                taskStatus = failWith(taskStatus, FAILED, failures);
+                if (cause instanceof TrinoTransportException) {
+                    // Since this TaskInfo is updated in the client the "complete" flag will not be set,
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:991]: What is the "complete" flag?

**Review by arhimondr — COMMENTED** [review]
Looks good to me. But I would prefer if somebody else could take a look as well. I've broken this code so many times that it feels like a second look wouldn't be unnecessary :smile:

File: core/trino-main/src/main/java/io/trino/operator/Driver.java [hunk 9]
```diff
@@ -526,6 +532,9 @@ private void destroyIfNecessary()
                     "Error destroying driver for task %s",
                     driverContext.getTaskId());
         }
+        finally {
+            destroyedFuture.set(null);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java:536]: So... I guess it's possible to move this outside of the locked region, but that will mean doing something like this in 3 other call sites:
```
finally {
  exclusiveLock.unlock();
  if (state.get() == DESTROYED) {
    destroyedFuture.set(null);
  }
}
```

Which seems more awkward to me in that it's spatially separating the action of destruction from setting the future. I Since right now that future is only tied to detecting termination at the task level, I think this is safe since the potential for deadlock is on the driver's exclusive lock and not on the any task-level lock- but it is riskier if the usage expands later on... maybe instead we don't expose `destroyedFuture` but instead allow passing a `Runnable` to the driver constructor directly? It's still the same thing as the cleanup action would run while holding the lock, but it seems less risky longer term since there's no opportunity to pack more listeners onto the driver later. Let me know what you think and I'll either keep what's there, move the future setting out of the lock region, or pass a cleanup runnable to the driver constructor instead.

File: core/trino-main/src/main/java/io/trino/operator/Driver.java [hunk 10]
```diff
@@ -183,15 +189,15 @@ public boolean isFinished()
...
+        return result.orElseGet(() -> state.get() != State.ALIVE || driverContext.isTerminatingOrDone());
     }
 
     @GuardedBy("exclusiveLock")
     private boolean isFinishedInternal()
     {
         checkLockHeld("Lock must be held to call isFinishedInternal");
 
-        boolean finished = state.get() != State.ALIVE || driverContext.isDone() || activeOperators.isEmpty() || activeOperators.get(activeOperators.size() - 1).isFinished();
+        boolean finished = state.get() != State.ALIVE || activeOperators.isEmpty() || activeOperators.get(activeOperators.size() - 1).isFinished() || driverContext.isTerminatingOrDone();
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java:200]: I'm ok with that, but the notion of "terminatingOr(Done,Finished)" doesn't apply super cleanly through all the usage sites. `Driver` has one specific definition (that's pretty complex in terms of what it entails), and `SplitRunner` has another that mostly overlaps, but also includes whether `DriverSplitRunner#close()` was called, and finally `PrioritizedSplitRunner` has another notion that includes whether `PrioritizedSplitRunner#destroy()` was called- so ultimately I think it's probably fine to keep the notion of `terminatingOrDone()` (task level concern) separate from `isFinished()` (`Driver`, `SplitRunner`, `PrioritizedSplitRunner` concern).

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java (hunk 7 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:846]: Nope, that's a mistake- will fix.

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java (hunk 8 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:991]: Haha, that's a great question. During the (fairly significant) refactoring here I was trying to preserve the original comments where possible, but you're right that actually there is indeed no "complete" flag anymore (was there ever?) so the comment is wrong. That does beg the question, _should_ there be a complete flag? I could add one to differentiate a "locally failed immediately" from "acknowledged termination" but no code would be using that at the moment.

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java (hunk 6 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:109]: It didn't create deadlocks, but it more tightly integrates the synchronized section to what's actually being guarded, which is synchronizing callers to `operatorFactories.createOperator()` and `operatorFactories.noMoreOperators()` (on each). The `checkState(!noMoreDrivers, ...)` call is always going to be prone to races between checking `DriverFactory#isNoMoreDrivers()` and `DriverFactory#createDriver()`.

So this _specific_ piece is just to constrain the locked region to the minimum part, but the other changes to make `isNoMoreDrivers()` not synchronize and `DriverSplitRunnerFactory#createDriver` handle the exception that could result from a race-prone check against `noMoreDrivers` instead of failing is more important.

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java [hunk 11]
```diff
@@ -71,6 +71,7 @@
 
     private Duration statusRefreshMaxWait = new Duration(1, TimeUnit.SECONDS);
     private Duration infoUpdateInterval = new Duration(3, TimeUnit.SECONDS);
+    private Duration taskTerminationTimeout = new Duration(10, TimeUnit.MINUTES);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:74]: Yeah, I didn't think too hard about what the right value should be because I wasn't sure how long we would expect this to take. Having tested this a bit, termination is pretty typically in the 10-20ms range so 1 minute sounds like a better default to me- I'll change it in the next revision.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 5 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:114]: Good idea, will refactor that into the tracker instead.

File: core/trino-main/src/main/java/io/trino/operator/Driver.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java:536]: Update on the above: you actually can’t move it out of the exclusive lock region without potentially missing the final DriverContext finished() call, since failing to acquire the lock could still see the state flip to DESTROYED before the lock holder completes destruction.

This could be addressed by adding a new DESTROYING state, but that’s extra complexity too.

File: core/trino-main/src/main/java/io/trino/operator/Driver.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java:200]: Yeah, I can see that. What do you think about changing it for the `Driver` only?

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java [hunk 12]
```diff
@@ -587,10 +620,33 @@ public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitioned
...
+                // driverFactory is already closed, ignore the exception and return null
+                if (driverFactory.isNoMoreDrivers()) {
+                    return null;
+                }
+                // this exception is unexpected if driverFactory has not been closed, so rethrow it
+                throw e;
+            }
 
+            // register driver destroyed listener to detect when termination completes
+            driver.getDestroyedFuture().addListener(driverAndTaskTerminationTracker, directExecutor());
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:649]: I'm not saying this is bad, but make sure you actually want direct executor here... depends on the on the context where destroyed future is set.

File: core/trino-main/src/main/java/io/trino/operator/Driver.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java:536]: Yeah, i can totally see how that would make the code less reasonable. I also agree that chances of introducing a deadlock here are rather low. However to me it still feels worth a decrease in readability, but I don't feel particularly strongly about it.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java [hunk 13]
```diff
@@ -816,4 +877,53 @@ public void stateChanged(BufferState newState)
             }
         }
     }
+
+    private static final class DriverAndTaskTerminationTracker
+            implements Runnable
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:882]: instead of making this `Runnable` use `driverAndTaskTerminationTracker::driverDestroyed` in the caller

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java [hunk 14]
```diff
@@ -816,4 +877,53 @@ public void stateChanged(BufferState newState)
...
+        }
+
+        public long getLiveCreatedDrivers()
+        {
+            return liveCreatedDrivers.get();
+        }
+
+        public void checkTaskTermination()
+        {
+            if (liveCreatedDrivers.get() == 0 && taskStateMachine.getState().isTerminating()) {
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:924]: With bugs this will go negative and the termination will never be called.  With the race in `tryCreateNewDriver` this could be positive when this is checked

File: core/trino-main/src/main/java/io/trino/operator/Driver.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java:200]: I think maybe the compromise is to rename `isFinishedInternal()` but not `Driver#isFinished()` since the internal method _is_ about checking whether the driver _should_ be destroyed (implies association to "terminating") but `Driver#isFinished()` is a separate semantic that needs only to describe having the Driver itself having "finished" (for whatever reason).

[the remark this answers is not in the document]
Comment (dain) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java]: Is there a reason this was reordered?  I'm just curious if the order matters here.  (if it does add a comment)

File: core/trino-main/src/main/java/io/trino/operator/Driver.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dain) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java]: I have the same concerns.  This code has a long history with deadlocks.  Maybe add a utility method like `checkDestroyedState`.
BTW a runnable won't work either as executing it while holding the lock is still an open out call.

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java [hunk 15]
```diff
@@ -141,7 +145,7 @@ public synchronized void noMoreDrivers()
         }
     }
 
-    public synchronized boolean isNoMoreDrivers()
+    public boolean isNoMoreDrivers()
     {
         return noMoreDrivers;
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:150]: I'm pretty sure you need an annotation to say that it is ok to ignore the guarded check here (and some comments on why)

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java [hunk 16]
```diff
@@ -38,7 +38,7 @@
     private final OptionalInt driverInstances;
 
     @GuardedBy("this")
-    private boolean noMoreDrivers;
+    private volatile boolean noMoreDrivers; // must synchronize between createDriver() and noMoreDrivers(), but isNoMoreDrivers() need not synchronize
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:41]: This change seems unrelated and should be moved to a separate commit

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java [hunk 17]
```diff
@@ -768,84 +782,112 @@ private synchronized SplitAssignment getSplitAssignment(PlanNodeId planNodeId, i
...
-            // send abort to task
-            scheduleAsyncCleanupRequest(new Backoff(maxErrorDuration), "abort", true);
+        else {
+            // check for termination timeout
+            long terminationStartedNanos = this.terminationStartedNanos.get();
+            if (terminationStartedNanos == 0) {
+                long currentTimeNanos = System.nanoTime();
+                // If reported time is exactly 0, increase it by 1 nanosecond so that "0" can be treated as "not set"
+                if (currentTimeNanos == 0) {
+                    currentTimeNanos++;
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:857]: use `currentTimeNanos = 1` instead

File: core/trino-main/src/main/java/io/trino/operator/Driver.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java]: After looking over some of the options, it seemed to me the best compromise between avoiding potential deadlocks and not having multiple checks in different codepaths to decide when to fire the `destroyedFuture` is to add `Driver.State.DESTROYING` to separate "destruction in progress by some thread" and "destruction completed", and have `DriverLock#unlock()` complete the future immediately after it's released- so I've published that version in the latest revision.

File: core/trino-main/src/main/java/io/trino/server/remotetask/RequestErrorTracker.java [hunk 18]
```diff
@@ -139,15 +137,14 @@ public void requestFailed(Throwable reason)
...
     }
 
-    @FormatMethod
     @SuppressWarnings("FormatStringAnnotation") // we manipulate the format string and there's no way to make Error Prone accept the result
-    static void logError(Throwable t, String format, Object... args)
+    static void logError(Throwable t, String message)
     {
         if (isExpectedError(t)) {
-            log.error(format + ": %s", ObjectArrays.concat(args, t));
+            log.error(message + ": " + t);
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/server/remotetask/RequestErrorTracker.java:144]:
````
log.error("%s: %s", message, t)

File: core/trino-main/src/main/java/io/trino/operator/Driver.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java]: > Is there a reason this was reordered? I'm just curious if the order matters here.

No correctness reason.

I reordered these checks originally while trying to disentangle race conditions and the isTerminatingOrDone() check used to eventually synchronize on TaskHandle.

At this point that check is now just a volatile read (landed in another PR, works similarly to how DriverFactory does in this PR), but I chose to keep the new ordering because it still holds that “normal” finishing is more likely than task termination and so the cheaper checks that of driver-local state should come first before trying to check shared state (or acquire shared locks).

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedQueryScheduler.java [hunk 19]
```diff
@@ -873,7 +874,8 @@ public static DistributedStagesScheduler create(
                 });
             }
 
-            Map<StageId, StageExecution> stageExecutions = new HashMap<>();
+            // Preserve topological ordering in stageExecutionsMap
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedQueryScheduler.java]: Because canceling a task will call `outputBuffer.destroy()`, which means that tasks consuming from said buffer can see `BufferResult(completed: true)` on the immediate subsequent request. This means that a task receiving a cancel before tasks consuming from it have themselves been canceled, aborted, or failed can trigger "finished" to propagate forward through the stages along worker to worker communication links unless the cancellation proceeds in topological order.

Essentially, when `DistributedStagesScheduler#cancel()` is called, we want to traverse the `stageExecutions` in topological order to reduce the possibility of that happening. That said, this issue shouldn't arise _now_ because `DistributedStagesScheduler#cancel()` is only called once the query state machine is finished anyway, but during development of these changes I did continue to see incorrect query results that seemed to stem from cancellations turning into incorrect query results, potentially due to tasks being sent both a cancel and abort command concurrently and those requests racing with one another to reach the to tasks (now fixed by `HttpRemoteTask` only allowing a single termination command to proceed).

Honestly, I'm still scratching my head a little bit about why tasks should call `outputBuffer.destroy()` on cancel instead of `outputBuffer.abort()`. It seems like that's a fragile behavior that can easily cause incorrect query results due to subtle concurrency or state machine bugs.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java [hunk 20]
```diff
@@ -587,10 +620,33 @@ public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitioned
...
-            Driver driver = driverFactory.createDriver(driverContext);
+            // Attempt to increment the driver count eagerly, but skip driver creation if the task is already terminating or done
+            if (!driverAndTaskTerminationTracker.tryCreateNewDriver()) {
+                return null;
+            }
+            Driver driver;
+            try {
+                driver = driverFactory.createDriver(driverContext);
+            }
+            catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:637]: True, will update this to catch `Throwable` instead but only suppress the exception and return null as long as the driver factory is closed and the caught exception is an `instanceof Exception` (to avoid suppressing fatal errors).

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 12 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:649]: For this particular case, I'm fairly confident that we want to use a direct executor. The action is short, only needs to perform an atomic update and maybe flip the state machine if termination has finished, and by doing it this way we know that we won't have a `RejectedExecutionException` causing tasks to get stuck and never finish terminating.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 14 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:924]: In the event of a race in `tryCreateNewDriver()` this could go positive for any particular check, but in that case the thread causing the race in `tryCreateNewDriver()` will notice the state machine is terminating and decrement the count back down and then re-check for termination.

I agree that if there's a bug and this goes negative then we're stuck but... what would we propose to do instead? Hard-fail the task state machine immediately even though drivers are still potentially running? Just consider anything negative as "successfully terminated"? I'm open to suggestions about what the best option is here.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 21]
```diff
@@ -250,15 +250,15 @@ public synchronized void cancel()
     {
         // Only send tasks a cancel command if the stage is successfully cancelled and not already failed
         if (stateMachine.transitionToCanceled()) {
-            getAllTasks().forEach(RemoteTask::cancel);
+            tasks.values().forEach(RemoteTask::cancel);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java]: This was peeled out and merged in another PR already, but it was indeed to avoid a redundant extra copy.

File: testing/trino-testing/src/main/java/io/trino/testing/QueryAssertions.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply testing/trino-testing/src/main/java/io/trino/testing/QueryAssertions.java:103]: I don't see a good way to get a `QueryId` out of `LocalQueryRunner` that would allow these code paths to be shared unfortunately, but if you see a way to do that let me know and I'll make that change. Agreed on the duplication here being less ideal.

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java (hunk 15 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:150]: You're right, the annotation is more correct over `operatorFactories` requiring the guarded access so I'll move the annotation there and leave the inline comment over `noMoreDrivers`.

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java (hunk 16 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:41]: Will do.

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java (hunk 17 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:857]: Done

File: core/trino-main/src/main/java/io/trino/server/remotetask/RequestErrorTracker.java (hunk 18 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/server/remotetask/RequestErrorTracker.java:144]: Good catch, done.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 13 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:882]: Done.

**Review by arhimondr — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedQueryScheduler.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (dain) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedQueryScheduler.java]: Sorry, I mean "explain why this matters in the comment"... so future readers know why sorting is there.  You don't need to explain in this level of detail (but you could).  Generally, I like to see enough to get the reader thinking of the implications.

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java [hunk 22]
```diff
@@ -33,12 +33,13 @@
...
     private final boolean outputDriver;
+    @GuardedBy("this")
     private final List<OperatorFactory> operatorFactories;
     private final Optional<PlanNodeId> sourceId;
     private final OptionalInt driverInstances;
 
-    @GuardedBy("this")
-    private boolean noMoreDrivers;
+    // must synchronize between createDriver() and noMoreDrivers(), but isNoMoreDrivers() is safe without synchronizing
+    private volatile boolean noMoreDrivers;
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:42]: I would leave the `@GuardedBy`, and add a supress warning (with comment) on `isNoMoreDrivers`

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 14 above)
Comment (dain) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:921]: Here is what I was thinking.  I don't know if this actually happens, but I think the code allows for this.
```
thread 1: tryCreateNewDriver incrementAndGet
thread 1: if isTerminatingOrDone is not entered
thread x: terminate task
thread 2: liveCreatedDrivers.get() will equal 1 and state is terminating is true, so terminationComplete is not called

Comment (dain) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:921]: For safety, I typically check `<= 0`, it mask over bugs (you could log), but it prevents bugs from locking up resources.  Also, calling check termination periodically after termination until the task is actually all cleaned up.  Again this can mask bugs, but it is better than locking up a server.

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java (hunk 22 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:42]: Ok, will change that in the latest revision.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 14 above)
Comment (dain) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java]: After off line discussion, my example above is covered by the call back handler added to the created driver.

**Review by dain — APPROVED** [review]
Looks good

**Comment by colebow** [comment]
A little late to the party, but it would be better to pick a term other than `ABORTING` if it can be helped.

**Comment by pettyjamesm** [comment]
> A little late to the party, but it would be better to pick a term other than `ABORTING` if it can be helped.

I’m open to renaming the state, but whatever new name we would choose should equally be applied to the “final” state name to maintain the symmetry of `FAILING` -> `FAILED`, `CANCELING` -> `CANCELED` and we’ve already got meaningful semantics that separate aborts from cancels so we’d need a new and different word.

> Also, does this need docs?

I can’t find any docs describing the existing task states (let me know if I missed something there), so I don’t think that there’s an immediate need to document the new ones although documenting task, stage, and query state semantics might be worthwhile to do at some point.

**Comment by mosabua** [comment]
One other request... ideally.. could we fix the grammar and change from CANCELING to CANCELLING ?

**Comment by mosabua** [comment]
And with regards to ... ABORTING .. could we change to TERMINATING ?

**Comment by pettyjamesm** [comment]
> And with regards to ... ABORTING .. could we change to TERMINATING ?

The problem I see with that is that failing, canceling, and aborting could all be considered “terminating” which is a useful concept by itself that unifies the three variations of “stopping the task early” that I’ve already used fairly heavily in this PR (e.g.: `TaskState#isTerminatingOrDone()`)

**Comment by mosabua** [comment]
Fair ... I think the current setup works. I just wish we could fix the spelling ;-)
